packages/dds/tree/src/util/rangeMap.ts

```diff
@@ -3,31 +3,94 @@
  * Licensed under the MIT License.
  */
 
-import { newTupleBTree, type TupleBTree } from "./bTreeUtils.js";
+import { assert } from "@fluidframework/core-utils/internal";
+import { BTree } from "@tylerbu/sorted-btree-es6";
 
 /**
- * RangeMap represents a mapping from integers to values of type T or undefined.
+ * RangeMap represents a mapping from keys of type K to values of type V or undefined.
+ * The set of all possible keys is assumed to be fully ordered,
+ * and for each key there should be a single next higher key.
  * The values for a range of consecutive keys can be changed or queried in a single operation.
+ * The structure of the keys is described by the `offsetKey` and `subtractKeys` functions provided in the constructor.
  */
-export class RangeMap<T> {
-	private readonly tree: TupleBTree<[number], RangeEntry<T>>;
+export class RangeMap<K, V> {
+	private tree: BTree<K, RangeEntry<V>>;
 
-	public constructor() {
-		this.tree = newTupleBTree();
+	/**
+	 * @param offsetKey - Function which returns a new key which is `offset` keys after `key`.
+	 * When `offset` is negative, the returned key should come before `key`.
+	 *
+	 * @param subtractKeys - Function which returns the difference between `b` and `a`.
+	 * Offsetting `b` by this difference should return `a`.
+	 * The difference can be infinite if `a` cannot be reached from `b` by offsetting,
+	 * but the difference should still be positive if `a` is larger than `b` and negative if smaller.
+	 */
+	public constructor(
+		private readonly offsetKey: (key: K, offset: number) => K,
+		private readonly subtractKeys: (a: K, b: K) => number,
+	) {
+		this.tree = new BTree(undefined, subtractKeys);
 	}
 
 	/**
 	 * Retrieves all entries from the RangeMap.
 	 */
-	public getAllEntries(): RangeQueryResult<T>[] {
-		const entries: RangeQueryResult<T>[] = [];
-		for (const [[start], entry] of this.tree.entries()) {
+	public entries(): RangeQueryEntry<K, V>[] {
+		const entries: RangeQueryEntry<K, V>[] = [];
+		for (const [start, entry] of this.tree.entries()) {
 			entries.push({ start, length: entry.length, value: entry.value });
 		}
 
 		return entries;
 	}
 
+	public clear(): void {
+		this.tree.clear();
+	}
+
+	public get(start: K, length: number): RangeQueryEntry<K, V>[] {
+		const result: RangeQueryEntry<K, V>[] = [];
+		const lastQueryKey = this.offsetKey(start, length - 1);
+
+		let nextKey = start;
+		let remainingLength = length;
+		{
+			const entry = this.tree.getPairOrNextLower(start);
+			if (entry !== undefined) {
+				const [key, { length: entryLength, value }] = entry;
+				const lastEntryKey = this.offsetKey(key, entryLength);
+				if (this.ge(lastEntryKey, start)) {
+					const overlappingLength = Math.min(
+						this.subtractKeys(lastEntryKey, start) + 1,
+						length,
+					);
+
+					result.push({ start, length: overlappingLength, value });
+					nextKey = this.offsetKey(lastEntryKey, 1);
```

Comment:
It may not be strictly necessary for correctness of this code (I have a hard time telling), but it seems like it would be good to update `remainingLength` here at least to avoid confusion.